Crash - java.lang.NullPointerException: at com.braintreepayments.cardform.CardScanningFragment.onActivityResult (CardScanningFragment.java:61)

### General information

* SDK/Library version: 3.3.0
* Android Version and Device: Samsung Galaxy S7 Edge (hero2qltespr), 4096MB RAM, Android 7.0

### Issue description

Hello, 

I saw this crash happening in my app. It has happen very few times so I don't think it's a big deal, but just thought I would let you know.

```
Caused by: java.lang.NullPointerException: 
  at com.braintreepayments.cardform.CardScanningFragment.onActivityResult (CardScanningFragment.java:61)
  at android.app.Activity.dispatchActivityResult (Activity.java:7250)
  at android.app.ActivityThread.deliverResults (ActivityThread.java:4521)
  at android.app.ActivityThread.performResumeActivity (ActivityThread.java:3806)
  at android.app.ActivityThread.handleResumeActivity (ActivityThread.java:3872)
  at android.app.ActivityThread.handleLaunchActivity (ActivityThread.java:3014)
  at android.app.ActivityThread.-wrap14 (ActivityThread.java)
  at android.app.ActivityThread$H.handleMessage (ActivityThread.java:1650)
  at android.os.Handler.dispatchMessage (Handler.java:102)
  at android.os.Looper.loop (Looper.java:154)
  at android.app.ActivityThread.main (ActivityThread.java:6688)
  at java.lang.reflect.Method.invoke (Method.java)
  at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run (ZygoteInit.java:1468)
  at com.android.internal.os.ZygoteInit.main (ZygoteInit.java:1358)
```

--- a/CardForm/src/main/java/com/braintreepayments/cardform/CardScanningFragment.java
+++ b/CardForm/src/main/java/com/braintreepayments/cardform/CardScanningFragment.java
@@ -4,6 +4,7 @@ import android.app.Activity;
 import android.app.Fragment;
 import android.content.Intent;
 import android.os.Bundle;
+import android.support.annotation.VisibleForTesting;
 
 import com.braintreepayments.cardform.utils.ColorUtils;
 import com.braintreepayments.cardform.view.CardForm;
@@ -13,11 +14,11 @@ import io.card.payment.CardIOActivity;
 public class CardScanningFragment extends Fragment {
 
     private static final int CARD_IO_REQUEST_CODE = 12398;
-    private static final String TAG = "com.braintreepayments.cardform.CardScanningFragment";
+    public static final String TAG = "com.braintreepayments.cardform.CardScanningFragment";
 
     private CardForm mCardForm;
 
-    public static void requestScan(Activity activity, CardForm cardForm) {
+    public static CardScanningFragment requestScan(Activity activity, CardForm cardForm) {
         CardScanningFragment fragment = (CardScanningFragment) activity.getFragmentManager().findFragmentByTag(TAG);
 
         if (fragment != null) {
@@ -34,12 +35,28 @@ public class CardScanningFragment extends Fragment {
                 .beginTransaction()
                 .add(fragment, TAG)
                 .commit();
+
+        return fragment;
+    }
+
+    public void setCardForm(CardForm cardForm) {
+        mCardForm = cardForm;
+    }
+
+    @Override
+    public void onSaveInstanceState(Bundle outState) {
+        super.onSaveInstanceState(outState);
+        outState.putBoolean("resuming", false);
     }
 
     @Override
     public void onCreate(Bundle savedInstanceState) {
         super.onCreate(savedInstanceState);
         setRetainInstance(true);
+
+        if (savedInstanceState != null && savedInstanceState.getBoolean("resuming")) {
+            return;
+        }
 
         Intent scanIntent = new Intent(getActivity(), CardIOActivity.class)
                 .putExtra(CardIOActivity.EXTRA_HIDE_CARDIO_LOGO, true)
@@ -58,7 +75,7 @@ public class CardScanningFragment extends Fragment {
     @Override
     public void onActivityResult(int requestCode, int resultCode, Intent data) {
         if (requestCode == CARD_IO_REQUEST_CODE) {
-            mCardForm.handleCardIOResponse(data);
+            mCardForm.handleCardIOResponse(resultCode, data);
 
             if (getActivity() != null) {
                 getActivity().getFragmentManager()

--- a/CardForm/src/main/java/com/braintreepayments/cardform/view/CardForm.java
+++ b/CardForm/src/main/java/com/braintreepayments/cardform/view/CardForm.java
@@ -3,12 +3,14 @@ package com.braintreepayments.cardform.view;
 import android.annotation.SuppressLint;
 import android.annotation.TargetApi;
 import android.app.Activity;
+import android.app.Fragment;
 import android.content.Context;
 import android.content.Intent;
 import android.os.Build.VERSION_CODES;
 import android.support.annotation.DrawableRes;
 import android.support.annotation.IntDef;
 import android.support.design.widget.TextInputEditText;
+import android.support.v4.app.FragmentManager;
 import android.text.Editable;
 import android.text.TextWatcher;
 import android.util.AttributeSet;
@@ -59,6 +61,7 @@ public class CardForm extends LinearLayout implements OnCardTypeChangedListener,
      * Shows the field, and require the field value to be non empty when validating the card form.
      */
     public static final int FIELD_REQUIRED = 2;
+    private CardScanningFragment mCardScanningFragment;
 
     /**
      * The statuses a field can be.
@@ -249,6 +252,13 @@ public class CardForm extends LinearLayout implements OnCardTypeChangedListener,
      * @param activity Used to set {@link android.view.WindowManager.LayoutParams#FLAG_SECURE} to prevent screenshots
      */
     public void setup(Activity activity) {
+        mCardScanningFragment = (CardScanningFragment) activity.getFragmentManager()
+                .findFragmentByTag(CardScanningFragment.TAG);
+
+        if (mCardScanningFragment != null) {
+            mCardScanningFragment.setCardForm(this);
+        }
+
         activity.getWindow().setFlags(WindowManager.LayoutParams.FLAG_SECURE,
                 WindowManager.LayoutParams.FLAG_SECURE);
 
@@ -341,13 +351,26 @@ public class CardForm extends LinearLayout implements OnCardTypeChangedListener,
      * @param activity
      */
     public void scanCard(Activity activity) {
-        if (isCardScanningAvailable()) {
-            CardScanningFragment.requestScan(activity, this);
+        if (isCardScanningAvailable() && mCardScanningFragment == null) {
+            mCardScanningFragment = CardScanningFragment.requestScan(activity, this);
         }
     }
 
+    /**
+     * Use {@link #handleCardIOResponse(int, Intent)} instead.
+     */
     @SuppressLint("DefaultLocale")
+    @Deprecated
     public void handleCardIOResponse(Intent data) {
+        handleCardIOResponse(Integer.MIN_VALUE, data);
+    }
+
+    @SuppressLint("DefaultLocale")
+    public void handleCardIOResponse(int resultCode, Intent data) {
+        if (resultCode == Activity.RESULT_CANCELED || resultCode == Activity.RESULT_OK) {
+            mCardScanningFragment = null;
+        }
+
         if (data != null && data.hasExtra(CardIOActivity.EXTRA_SCAN_RESULT)) {
             CreditCard scanResult = data.getParcelableExtra(CardIOActivity.EXTRA_SCAN_RESULT);
 

--- a/CardForm/src/test/java/com/braintreepayments/cardform/view/CardFormTest.java
+++ b/CardForm/src/test/java/com/braintreepayments/cardform/view/CardFormTest.java
@@ -15,6 +15,7 @@ import android.widget.EditText;
 import android.widget.ImageView;
 import android.widget.TextView;
 
+import com.braintreepayments.cardform.CardScanningFragment;
 import com.braintreepayments.cardform.OnCardFormFieldFocusedListener;
 import com.braintreepayments.cardform.OnCardFormValidListener;
 import com.braintreepayments.cardform.R;
@@ -29,6 +30,7 @@ import org.robolectric.RobolectricTestRunner;
 import org.robolectric.RuntimeEnvironment;
 import org.robolectric.android.controller.ActivityController;
 
+import java.lang.reflect.Field;
 import java.util.concurrent.atomic.AtomicBoolean;
 import java.util.concurrent.atomic.AtomicInteger;
 
@@ -47,6 +49,8 @@ import static com.braintreepayments.cardform.test.TestExpirationDate.getValidExp
 import static junit.framework.Assert.assertEquals;
 import static junit.framework.Assert.assertFalse;
 import static junit.framework.Assert.assertTrue;
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertNull;
 import static org.mockito.Mockito.mock;
 import static org.mockito.Mockito.spy;
 import static org.mockito.Mockito.verify;
@@ -1850,7 +1854,7 @@ public class CardFormTest {
                 .mobileNumberExplanation("Make sure SMS is supported")
                 .setup(mActivity);
 
-        mCardForm.handleCardIOResponse(null);
+        mCardForm.handleCardIOResponse(Integer.MIN_VALUE, null);
 
         assertEquals("", ((EditText) mActivity.findViewById(R.id.bt_card_form_card_number)).getText().toString());
         assertEquals("", ((EditText) mActivity.findViewById(R.id.bt_card_form_expiration)).getText().toString());
@@ -1858,6 +1862,68 @@ public class CardFormTest {
         assertEquals("", ((EditText) mActivity.findViewById(R.id.bt_card_form_cardholder_name)).getText().toString());
         assertEquals("", ((EditText) mActivity.findViewById(R.id.bt_card_form_postal_code)).getText().toString());
         assertEquals("", ((EditText) mActivity.findViewById(R.id.bt_card_form_mobile_number)).getText().toString());
+    }
+
+    @Test
+    public void handleCardIOResponse_whenResultCancelled_clearsFragment() throws Exception {
+        mCardForm.cardRequired(true)
+                .expirationRequired(true)
+                .cvvRequired(true)
+                .cardholderName(CardForm.FIELD_REQUIRED)
+                .postalCodeRequired(true)
+                .mobileNumberRequired(true)
+                .mobileNumberExplanation("Make sure SMS is supported")
+                .setup(mActivity);
+
+        Field mCardScanningFragmentField = CardForm.class
+                .getDeclaredField("mCardScanningFragment");
+        mCardScanningFragmentField.setAccessible(true);
+        mCardScanningFragmentField.set(mCardForm, new CardScanningFragment());
+
+        mCardForm.handleCardIOResponse(Activity.RESULT_CANCELED, null);
+
+        assertNull(mCardScanningFragmentField.get(mCardForm));
+    }
+
+    @Test
+    public void handleCardIOResponse_whenResultOK_clearsFragment() throws Exception {
+        mCardForm.cardRequired(true)
+                .expirationRequired(true)
+                .cvvRequired(true)
+                .cardholderName(CardForm.FIELD_REQUIRED)
+                .postalCodeRequired(true)
+                .mobileNumberRequired(true)
+                .mobileNumberExplanation("Make sure SMS is supported")
+                .setup(mActivity);
+
+        Field mCardScanningFragmentField = CardForm.class
+                .getDeclaredField("mCardScanningFragment");
+        mCardScanningFragmentField.setAccessible(true);
+        mCardScanningFragmentField.set(mCardForm, new CardScanningFragment());
+        mCardForm.handleCardIOResponse(Activity.RESULT_OK, null);
+
+        assertNull(mCardScanningFragmentField.get(mCardForm));
+    }
+
+    @Test
+    public void handleCardIOResponse_whenResultNotOkOrCancelled_retainsFragment() throws Exception {
+        mCardForm.cardRequired(true)
+                .expirationRequired(true)
+                .cvvRequired(true)
+                .cardholderName(CardForm.FIELD_REQUIRED)
+                .postalCodeRequired(true)
+                .mobileNumberRequired(true)
+                .mobileNumberExplanation("Make sure SMS is supported")
+                .setup(mActivity);
+
+        Field mCardScanningFragmentField = CardForm.class
+                .getDeclaredField("mCardScanningFragment");
+        mCardScanningFragmentField.setAccessible(true);
+        mCardScanningFragmentField.set(mCardForm, new CardScanningFragment());
+
+        mCardForm.handleCardIOResponse(Activity.RESULT_FIRST_USER, null);
+
+        assertNotNull(mCardScanningFragmentField.get(mCardForm));
     }
 
     @Test
@@ -1871,7 +1937,7 @@ public class CardFormTest {
                 .mobileNumberExplanation("Make sure SMS is supported")
                 .setup(mActivity);
 
-        mCardForm.handleCardIOResponse(new Intent());
+        mCardForm.handleCardIOResponse(Integer.MIN_VALUE, new Intent());
 
         assertEquals("", ((EditText) mActivity.findViewById(R.id.bt_card_form_card_number)).getText().toString());
         assertEquals("", ((EditText) mActivity.findViewById(R.id.bt_card_form_expiration)).getText().toString());
@@ -1894,7 +1960,7 @@ public class CardFormTest {
         Intent intent = new Intent()
                 .putExtra(CardIOActivity.EXTRA_SCAN_RESULT, new CreditCard(VISA, 0, 0, "", "", ""));
 
-        mCardForm.handleCardIOResponse(intent);
+        mCardForm.handleCardIOResponse(Activity.RESULT_OK, intent);
 
         assertEquals(VISA, ((EditText) mActivity.findViewById(R.id.bt_card_form_card_number)).getText().toString());
         assertTrue(mCardForm.findViewById(R.id.bt_card_form_expiration).hasFocus());
@@ -1913,7 +1979,7 @@ public class CardFormTest {
         Intent intent = new Intent()
                 .putExtra(CardIOActivity.EXTRA_SCAN_RESULT, new CreditCard(VISA, 0, 0, "", "", ""));
 
-        mCardForm.handleCardIOResponse(intent);
+        mCardForm.handleCardIOResponse(Activity.RESULT_OK, intent);
 
         assertEquals("", ((EditText) mActivity.findViewById(R.id.bt_card_form_card_number)).getText().toString());
     }
@@ -1931,7 +1997,7 @@ public class CardFormTest {
         Intent intent = new Intent()
                 .putExtra(CardIOActivity.EXTRA_SCAN_RESULT, new CreditCard(VISA, 12, 2020, "", "", ""));
 
-        mCardForm.handleCardIOResponse(intent);
+        mCardForm.handleCardIOResponse(Activity.RESULT_OK, intent);
 
         assertEquals("122020", ((EditText) mActivity.findViewById(R.id.bt_card_form_expiration)).getText().toString());
         assertTrue(mCardForm.findViewById(R.id.bt_card_form_cvv).hasFocus());
@@ -1950,7 +2016,7 @@ public class CardFormTest {
         Intent intent = new Intent()
                 .putExtra(CardIOActivity.EXTRA_SCAN_RESULT, new CreditCard(VISA, 7, 2020, "", "", ""));
 
-        mCardForm.handleCardIOResponse(intent);
+        mCardForm.handleCardIOResponse(Activity.RESULT_OK, intent);
 
         assertEquals("07", ((ExpirationDateEditText) mActivity.findViewById(R.id.bt_card_form_expiration)).getMonth());
         assertEquals("2020", ((ExpirationDateEditText) mActivity.findViewById(R.id.bt_card_form_expiration)).getYear());
@@ -1970,7 +2036,7 @@ public class CardFormTest {
         Intent intent = new Intent()
                 .putExtra(CardIOActivity.EXTRA_SCAN_RESULT, new CreditCard(VISA, 12, 2000, "", "", ""));
 
-        mCardForm.handleCardIOResponse(intent);
+        mCardForm.handleCardIOResponse(Activity.RESULT_OK, intent);
 
         assertEquals("", ((EditText) mActivity.findViewById(R.id.bt_card_form_expiration)).getText().toString());
         assertTrue(mCardForm.findViewById(R.id.bt_card_form_expiration).hasFocus());
@@ -1989,7 +2055,7 @@ public class CardFormTest {
         Intent intent = new Intent()
                 .putExtra(CardIOActivity.EXTRA_SCAN_RESULT, new CreditCard(VISA, 12, 2020, "", "", ""));
 
-        mCardForm.handleCardIOResponse(intent);
+        mCardForm.handleCardIOResponse(Activity.RESULT_OK, intent);
 
         assertEquals("", ((EditText) mActivity.findViewById(R.id.bt_card_form_expiration)).getText().toString());
         assertTrue(mCardForm.findViewById(R.id.bt_card_form_cvv).hasFocus());

